PhysicsTools/PatUtils/interface/SmearedJetProducerT.h

```diff
@@ -21,370 +17,256 @@
-     ~GenJetMatcherT()
-     {
-       delete dRmaxGenJetMatch_;
-     }
-
-     const reco::GenJet* operator()(const T& jet, edm::Event* evt = 0) const
-     {
-       assert(evt);
-
-       edm::Handle<reco::GenJetCollection> genJets;
-       evt->getByToken(srcGenJetsToken_, genJets);
-
-       const reco::GenJet* retVal = 0;
-
-       double dRbestMatch = 1.e+6;
-       for ( reco::GenJetCollection::const_iterator genJet = genJets->begin();
-	     genJet != genJets->end(); ++genJet ) {
-	 double dRmax = dRmaxGenJetMatch_->Eval(genJet->pt());
-	 //std::cout << "genJetPt = " << genJet->pt() << ": dRmax = " << dRmax << std::endl;
-	 double dR = deltaR(jet.p4(), genJet->p4());
-	 if ( dR < dRbestMatch && dR < dRmax ) {
-	   retVal = &(*genJet);
-	   dRbestMatch = dR;
-	 }
-       }
-
-       return retVal;
-     }
+            m_jets_token = consumes<JetCollection>(cfg.getParameter<edm::InputTag>("src"));
 
-    private:
+            if (m_enabled) {
+                m_rho_token = consumes<double>(cfg.getParameter<edm::InputTag>("rho"));
 
-//--- configuration parameter
-     edm::InputTag srcGenJets_;
-     edm::EDGetTokenT<reco::GenJetCollection> srcGenJetsToken_;
+                m_use_txt_files = cfg.exists("resolutionFile") && cfg.exists("scaleFactorFile");
 
-     TFormula* dRmaxGenJetMatch_;
-  };
+                if (m_use_txt_files) {
+                    std::string resolutionFile = cfg.getParameter<edm::FileInPath>("resolutionFile").fullPath();
+                    std::string scaleFactorFile = cfg.getParameter<edm::FileInPath>("scaleFactorFile").fullPath();
 
-  template <typename T>
-  class JetResolutionExtractorT
-  {
-    public:
+                    m_resolution_from_file.reset(new JME::JetResolution(resolutionFile));
+                    m_scale_factor_from_file.reset(new JME::JetResolutionScaleFactor(scaleFactorFile));
+                } else {
+                    m_jets_algo = cfg.getParameter<std::string>("algo");
+                }
 
-     JetResolutionExtractorT(const edm::ParameterSet&) {}
-     ~JetResolutionExtractorT() {}
+                std::uint32_t seed = cfg.getUntrackedParameter<std::uint32_t>("seed", 37428479);
+                m_random_generator = std::mt19937(seed);
 
-     double operator()(const T&) const
-     {
-       throw cms::Exception("SmearedJetProducer::produce")
-	 << " Jets of type other than PF not supported yet !!\n";
-     }
-  };
+                bool skipGenMatching = cfg.getUntrackedParameter<bool>("skipGenMatching", false);
+                if (! skipGenMatching)
+                    m_genJetMatcher = std::make_shared<pat::GenJetMatcher>(cfg, consumesCollector());
 
-  template <typename T>
-  class RawJetExtractorT // this template is neccessary to support pat::Jets
-                         // (because pat::Jet->p4() returns the JES corrected, not the raw, jet momentum)
-  {
-    public:
+                std::int32_t variation = cfg.getUntrackedParameter<std::int32_t>("variation", 0);
```

Comment:
It does indeed. I've put them untracked because I'd like to have default parameters, which is not possible (AFAIK) if the parameters are tracked. Do you have an alternative?